paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala

```diff
@@ -49,10 +51,25 @@ case class WriteIntoPaimonTable(
   private lazy val mergeSchema = options.get(SparkConnectorOptions.MERGE_SCHEMA)
 
   override def run(sparkSession: SparkSession): Seq[Row] = {
+    var data = _data
     if (mergeSchema) {
       val dataSchema = SparkSystemColumns.filterSparkSystemColumns(data.schema)
       val allowExplicitCast = options.get(SparkConnectorOptions.EXPLICIT_CAST)
       mergeAndCommitSchema(dataSchema, allowExplicitCast)
+
+      // For case that some columns is absent in data, we still allow to write once write.merge-schema is true.
+      val newTableSchema = SparkTypeUtils.fromPaimonRowType(table.schema().logicalRowType())
+      if (!schemaEqualsIgnoreNullability(newTableSchema, data.schema)) {
```

Comment:
Maybe can refer to spark `DataType.sameType` or `DataType.equalsIgnoreNullability`